### PR TITLE
Add CBMC proof for vProcessGeneratedUDPPacket_IPv6

### PR DESCRIPTION
**Pull request by moninom1**
Description
-----------
Add CBMC proof for vProcessGeneratedUDPPacket_IPv6
It  also validates : 

FreeRTOS_FindEndPointOnNetMask
FreeRTOS_InterfaceEndPointOnNetMask
prvStartLookup

Test Steps
-----------
CBMC test 
```
python prepare.py
make
```

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by tony-josi-aws — DISMISSED** [review]

**Review by ActoryOu — APPROVED** [review]

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/Makefile.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
+  "ENTRY": "vProcessGeneratedUDPPacket_IPv6",
+  "CBMCFLAGS":
+  [
```
Comment (AniruddhaKanhere) [inline test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/Makefile.json:4]: No unwinding flag?

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+
+/* We do not need to calculate the actual checksum for the proof to be complete.
+ * Neither does the checksum matter for completeness. */
+uint16_t usGenerateProtocolChecksum( const uint8_t * const pucEthernetBuffer,
+                                     size_t uxBufferLength,
+                                     BaseType_t xOutgoingPacket )
+{
+    __CPROVER_assert( pucEthernetBuffer != NULL, "The Ethernet buffer cannot be NULL while generating Protocol Checksum" );
```
Comment (AniruddhaKanhere) [inline test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c:43]: Should we assert on the ethernet header being non-NULL as well?

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    eARPLookupResult_t eResult;
+
+    return eResult;
+}
+
+/* This function has been tested separately. Therefore, we assume that the implementation is correct. */
+void vNDSendNeighbourSolicitation( NetworkBufferDescriptor_t * pxNetworkBuffer,
+                                   const IPv6_Address_t * pxIPAddress )
+{
+    __CPROVER_assert( pxNetworkBuffer != NULL, "The network buffer descriptor cannot be NULL." );
```
Comment (AniruddhaKanhere) [inline test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c:67]: Same here, should we be asserting that the ethernet buffer in the network buffer is also non-NULL?

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    /* The buffer cannot be NULL for the function call. */
+    __CPROVER_assume( pxNetworkBuffer != NULL );
+    __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer != NULL );
+
+    /*
+     * Add an end point to the network buffer present. Its assumed that the
+     * network interface layer correctly assigns the end point to the generated buffer.
+     */
+    pxNetworkBuffer->pxEndPoint = ( NetworkEndPoint_t * ) malloc( sizeof( NetworkEndPoint_t ) );
+    __CPROVER_assume( pxNetworkBuffer->pxEndPoint != NULL );
```
Comment (AniruddhaKanhere) [inline test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c:103]: malloc will never fail so why do we need to assume?
I think you might have meant to use safeMalloc (or use the `--malloc-may-fail` and the `--malloc-fail-null` flags).

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    /* Add few endpoints to global pxNetworkEndPoints */
+    pxNetworkEndPoints = ( NetworkEndPoint_t * ) malloc( sizeof( NetworkEndPoint_t ) );
+    __CPROVER_assume( pxNetworkEndPoints != NULL );
+
+    pxNetworkEndPoints->pxNetworkInterface = pxNetworkBuffer->pxEndPoint->pxNetworkInterface;
+
+    if( nondet_bool() )
+    {
+        pxNetworkEndPoints->pxNext = ( NetworkEndPoint_t * ) malloc( sizeof( NetworkEndPoint_t ) );
+        __CPROVER_assume( pxNetworkEndPoints->pxNext != NULL );
```
Comment (AniruddhaKanhere) [inline test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c:119]: Same here.

**Review by AniruddhaKanhere — COMMENTED** [review]
Thank you @moninom1 for this PR. A few comments on this one. Can you please take a look?

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c (hunk 4 above)
Comment (moninom1) [reply test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c:103]: Yes we are using safeMalloc everywhere for covering both scenerio, i think i missed at these places, will update.

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c (hunk 5 above)
Comment (moninom1) [reply test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c:119]: Will update to safeMalloc

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/Makefile.json (hunk 1 above)
Comment (moninom1) [reply test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/Makefile.json:4]: There was no loop.

File: test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c (hunk 2 above)
Comment (moninom1) [reply test/cbmc/proofs/UDP/vProcessGeneratedUDPPacket_IPv6/vProcessGeneratedUDPPacket_IPv6_harness.c:43]: I see that assert is there in the usGenerateProtocolChecksum as well. So for generalization if there is assert in main function we should do __CPROVER_assert for that?

**Review by shubnil — DISMISSED** [review]

**Review by tony-josi-aws — APPROVED** [review]

**Review by shubnil — APPROVED** [review]

**Pull request merged**